src/core/thread/mle_router.cpp

```diff
@@ -238,7 +238,7 @@ Error MleRouter::BecomeRouter(ThreadStatusTlv::Status aStatus)
     Error error = kErrorNone;
 
     VerifyOrExit(!IsDisabled(), error = kErrorInvalidState);
-    VerifyOrExit(!IsRouter(), error = kErrorNone);
+    VerifyOrExit(!IsRouter() && !IsLeader(), error = kErrorNone);
```

Comment:
May be good to check that adding `IsLeader()` is safe/fine here, i.e. we do not call `BecomeRouter()` anywhere in the code when the current role may be "leader".

Reply:
I did go through all places in the OT core code where `BecomeRouter()` is called, seems to be okay to add `IsLeader()` check. 

We also expose this as a public API, but it is intended for testing, so changing its behavior should  be fine.